# A crash course on Python for Data Analysis

1.	Python basic programming:
    + Python modules or how to extend the instruction set
    + Data Types and Operations.
    + A program in Python
    + A function in Python
    + Slicing
    + References
    + Comprehensions
    + Generators
    + Objects
    + Reading Data

## Python

The software program that you use to invoke operators is called an **interpreter**. You enter your commands as a ‘dialog’ between you and the interpreter. Commands can be entered as part of a script (a text file with a list of commands to perform) or directly at the *cell*. 

In order to ask to the interpreter what to do, you must **invoke** an operator:

In [1]:
!python --version

Python 3.6.7


In [2]:
3 + 4 + 9

16

In [5]:
list(range(10))

[0, 1, 2, 3, 4, 5, 6, 7, 8, 9]

It’s helpful to think of the computation carried out by an operator as involving four parts:

+ The name of the operator
+ The input arguments
+ The output value
+ Side effects

A typical operation takes one or more input arguments and uses the information in these to produce an output value. Along the way, the computer might take some action: display a graph, store a file, make a sound, etc. These actions are called side effects.

### Modules

Python is a general-purpose programming language, so when we want to use more specific commands (such as statistical operators or string processing operators) we usually need to import them before we can use them. 

For Scientific Python, one of the most important libraries that we need is **numpy** (Numerical Python), which can be loaded like this:

In [6]:
import numpy as np
np.sqrt(25)

5.0

In [7]:
np.arange(10)

array([0, 1, 2, 3, 4, 5, 6, 7, 8, 9])

Access to the functions, variables and classes of a module depends on the way the module was imported:

In [8]:
import math
math.cos(math.pi)

-1.0

In [9]:
import math as m  # import using an alias
m.cos(m.pi)

-1.0

In [10]:
from math import cos,pi # import only some functions
cos(pi)

-1.0

In [11]:
from math import *   # global import
cos(pi)

-1.0

### Objects

Often the value returned by an operation will be used later on. Values can be stored for later use with the **assignment operator**:

In [0]:
a = 101

The command has stored the value 101 under the name <code>a</code>. Such stored values are called **objects**. 

Making an assignment to an object defines the object. Once an object has been defined, it can be referred to and used in later computations. 

To refer to the value stored in the object, just use the object’s name itself. For instance:

In [13]:
b = np.sqrt(a)
b

10.04987562112089

There are some general rules for object names:

+ Use only letters and numbers and ‘underscores’ (_)
+ Do NOT use spaces anywhere in the name
+ A number cannot be the first character in the name
+ Capital letters are treated as distinct from lower-case letters (i.e., Python is case-sensitive)

In [14]:
3a = 10

SyntaxError: ignored

### Dynamic Typing

When you assign a new value to an existing object (*dynamic typing*), the former values of that object is erased from the computer memory. The former value of b was 10.0498756211, but after a new assignment:

In [16]:
b = 'a'
print(b)

a


The value of an object is changed only via the assignment operator. Using an object in a computation does not change the value. 

The brilliant thing about organizing operators in terms of input arguments and output values is that the output of one operator can be used as an input to another. This lets complicated computations be built out of simpler ones.

One way to connect the computations is by using objects to store the intermediate outputs:

In [17]:
a = np.arange(5)
np.sqrt(a)

array([0.        , 1.        , 1.41421356, 1.73205081, 2.        ])

You can also pass the output of an operator directly as an argument to another operator:

In [18]:
np.sqrt(np.arange(5))

array([0.        , 1.        , 1.41421356, 1.73205081, 2.        ])

### Data Types

Most of the examples used so far have dealt with numbers. But computers work with other kinds of information as well: text, photographs, sounds, sets of data, and so on. The word *type* is used to refer to the kind of information. 

It’s important to know about the types of data because operators expect their input arguments to be of specific types. When you use the wrong type of input, the computer might not be able to process your command.

For our purposes, it’s important to distinguish among several basic types:

+ Numeric (positive and negative) data: 
    + decimal and fractional numbers (**floats**), <code>a = 3.5</code>
    + whole numbers (**integers**), <code> b = -12560</code>, and 
    + arbitrary length whole numbers (**longs**):  <code>c=1809109863596239561236235625629561L</code>
+ **Strings** of textual data - you indicate string data to the computer by enclosing the text in quotation marks (e.g., <code>name = "python"</code>).
+ **Boolean** data: <code>a = True</code> or <code>a = False</code>.
+ **Complex** numbers: <code>a = 2+3j</code>
+ Sequence types: **tuples, lists, sets, dictionaries** and **files**.

In [0]:
a = 10

In [21]:
a = 'a'
print(a)

a


### Operators

+ Addition (also string, tuple and list concatenation) <code>a + b</code>
+ Subtraction (also set difference): <code>a - b</code>
+ Multiplication (also string, tuple and list replication): <code>a * b</code>
+ Division: <code>a / b</code>
+ Truncated integer division (rounded towards minus infinity): <code>a // b</code>
+ Modulus or remainder: <code>a % b</code>
+ Exponentiation: <code>a ** b</code>
+ Assignment: <code>=</code>, <code>-=</code>, <code>+=</code>,<code>/=</code>,<code>*=</code>, <code>%=</code>, <code>//=</code>, <code>**=</code>
+ Boolean comparisons: <code>==</code>, <code>!=</code>, <code><</code>,<code>></code>,<code><=</code>, <code>>=</code>
+ Boolean operators: <code>and</code>, <code>or</code>, <code>not</code>
+ Membership test operators: <code>in</code>, <code>not in</code>
+ Object identity operators: <code>is</code>, <code>is not</code>
+ Bitwise operators (or, xor, and, complement): <code>|</code>, <code>^</code>, <code>&</code>, <code>~</code>
+ Left and right bit shift: <code><<</code>, <code>>></code>

### Python as a calculator

The Python language has a concise notation for arithmetic that looks very much like the traditional one.

In [23]:
a = 3+2
b= 3.5 * -8
c = 10/6
print(a, b, c, 10./6.)

5 -28.0 1.6666666666666667 1.6666666666666667


Some math functions are not available in the basic Python module, and they need to be imported from a specific module:

In [25]:
import math   # this instruction is not executed if the module has already been imported
print(math.pi + math.sin(100) + math.ceil(2.3))

5.635227012480034


### A program in Python

General Rules:

+ All text from a <code>#</code> simbol to the end of a line are considered as comments.
+ Code must be **indented** and sometimes delineated by colons. The Python standard for indentation is four spaces. Never use tabs: it can produce hard to find errors. Set you editor to convert tabs to spaces.
+ Typically, a statement must be on a line. You can use a backslash <code>\</code> at the end of a line to continue a statement on to the next line.


In [28]:
# This program computes the factorial of 100.

fact = 1
n= 100
for factor in range(n,0,-1):
    fact = fact * \
    factor 
print(fact)    

93326215443944152681699238856266700490715968264381621468592963895217599993229915608941463976156518286253697920827223758251185210916864000000000000000000000000


In [29]:
list(range(10,0,-1))

[10, 9, 8, 7, 6, 5, 4, 3, 2, 1]

<div class="alert alert-error"> When we write a colon at the end of an iteration, all lines indented at the next level are considered *part* of the iteration. 

When we write a line at the same indentation as the iteration, we are closing the iteration.</div>

### A function in Python



#### Factorial

The factorial of a non-negative integer $n$, denoted by $n!$, is the product of all positive integers less than or equal to $n$.  

In [0]:
def factorial(n):
    fact = 1
    for factor in range(n,0,-1):
        fact = fact * factor
    return fact

In [32]:
factorial(100)

93326215443944152681699238856266700490715968264381621468592963895217599993229915608941463976156518286253697920827223758251185210916864000000000000000000000000

#### Fibonacci

The Fibonacci Sequence is the series of numbers: 0, 1, 1, 2, 3, 5, 8, 13, 21, 34, ...

The general rule to compute the sequence is very simple: The next number is found by adding up the two numbers before it.

In [34]:
def fib1(n):
    if n==1:
        return 1
    if n==0:
        return 0
    return fib1(n-1) + fib1(n-2)

fib1(20)

# this function cannot compute fib(100)

6765

In [35]:
import this

The Zen of Python, by Tim Peters

Beautiful is better than ugly.
Explicit is better than implicit.
Simple is better than complex.
Complex is better than complicated.
Flat is better than nested.
Sparse is better than dense.
Readability counts.
Special cases aren't special enough to break the rules.
Although practicality beats purity.
Errors should never pass silently.
Unless explicitly silenced.
In the face of ambiguity, refuse the temptation to guess.
There should be one-- and preferably only one --obvious way to do it.
Although that way may not be obvious at first unless you're Dutch.
Now is better than never.
Although never is often better than *right* now.
If the implementation is hard to explain, it's a bad idea.
If the implementation is easy to explain, it may be a good idea.
Namespaces are one honking great idea -- let's do more of those!


In [37]:
def fib2(n):
    a, b = 0, 1
    for i in range(1,n+1):
        a, b = b, a + b
    return a

n = 1000

if n<15:
    print(fib1(n))
else: 
    print(fib2(n))

43466557686937456435688527675040625802564660517371780402481729089536555417949051890403879840079255169295922593080322634775209689623239873322471161642996440906533187938298969649928516003704476137795166849228875


#### Greatest Common Divisor

The greatest common divisor of two positive integers $a$ and $b$ is the largest divisor common to $a$ and $b$.  

The Euclidean algorithm, or Euclid's algorithm, is an interative method for computing the greatest common divisor of two integers. 

+ If $a<b$, exchange $a$ and $b$.
+ Divide $a$ by $b$ and get the remainder, $r$. If $r=0$, report $b$ as the GCD of $a$ and $b$.
+ Replace $a$ by $b$ and replace $b$ by $r$. If $r \neq 0$ iterate.



In [38]:
def gcd(a,b): # Euclides algorithm v1.0: pseudocode translation
    r = 1
    while r != 0:
        if a<b:
            c=a
            a=b
            b=c
        r = a%b 
        if r == 0:
            return b
        else:
            a = b
            b = r

gcd(100,16)

4

In [39]:
def gcd(a,b):   # Euclides algorithm v2.0: idiomatic Python
    while a:
        a, b = b%a, a
    return b

gcd(100,16)

4

In [40]:
a = 0
a == False

True

### String processing with Python

Strings are list of characters:

In [41]:
a = 'python'
type(a)

str

In [43]:
print("Hello")

Hello


In [44]:
print("This is 'an example' of the use of quotes and double quotes")
print('This is "another example" of the use of quotes and double quotes')

This is 'an example' of the use of quotes and double quotes
This is "another example" of the use of quotes and double quotes


We can use the operator ``+`` to concatenate strings:

In [46]:
a = 'He'
b = 'llo'
c = a+b+'!'
print(c)

Hello!


Substrings within a string can be accessed using **slicing**. Slicing uses ``[]`` to contain the indices of the characters in a string, where the first index is $0$, and the last is $n - 1$ (assuming the string has $n$ characters). 

In [48]:
a = 'Python'
print(a[:], a[1], a[2:], a[:3], a[2:4], a[::2], a[1::2])

Python y thon Pyt th Pto yhn


The most advanced string functiona are stored in an external module called ``string``

In [49]:
import string as st
help(st)

Help on module string:

NAME
    string - A collection of string constants.

MODULE REFERENCE
    https://docs.python.org/3.6/library/string
    
    The following documentation is automatically generated from the Python
    source files.  It may be incomplete, incorrect or include features that
    are considered implementation detail and may vary between Python
    implementations.  When in doubt, consult the module reference at the
    location listed above.

DESCRIPTION
    Public module variables:
    
    whitespace -- a string containing all ASCII whitespace
    ascii_lowercase -- a string containing all ASCII lowercase letters
    ascii_uppercase -- a string containing all ASCII uppercase letters
    ascii_letters -- a string containing all ASCII letters
    digits -- a string containing all ASCII decimal digits
    hexdigits -- a string containing all ASCII hexadecimal digits
    octdigits -- a string containing all ASCII octal digits
    punctuation -- a string containing all

In [51]:
type(float('10.3'))

float

In [0]:
a = 'a'

In [0]:
# press tab for help after .
a.

In [54]:
a='Hello'
b = a.lower()
print(b)

hello


In [56]:
print(st.ascii_letters)
'a' in st.ascii_letters

abcdefghijklmnopqrstuvwxyzABCDEFGHIJKLMNOPQRSTUVWXYZ


True

### Lists

Lists are a built-in data type which require other data types to be useful. A list is a collection of other objects – floats, integers, complex numbers, strings or even other lists.

Lists also support slicing to retrieve one or more elements. Basic lists are constructed using square braces, ``[]``, and values are separated using
commas.

In [57]:
l=[]
type(l)

list

In [59]:
x=[1,2,3,4,[1,2,3,4],'jordi']
print(x[4:], x[0], x[5])

[[1, 2, 3, 4], 'jordi'] 1 jordi


In [60]:
x[-2:]  # The stride can also be negative which can be used to select the
        # elements of a list in reverse order.

[[1, 2, 3, 4], 'jordi']

Lists can be multidimensional and slicing can be done directly in higher dimensions:

In [61]:
x = [[1,2,3,4], [5,6,7,8]]
print(x[0], x[0][0])

[1, 2, 3, 4] 1


### Conditionals

The conditional structure in Python is <code>If</code>. It is usally combined with 
relational operators: <code> <, <=, ==, >=, >, != </code>.

In [65]:
def main(celsius):
    fahrenheit = 9.0 /5.0 * celsius + 32
    print("The temperature in Fahrenheit is", fahrenheit)
    if fahrenheit > 90:
        print("It's really hot out there.")
    elif fahrenheit < 30:
        print("It's really cold out there.")
    else: pass
        
main(35)

The temperature in Fahrenheit is 95.0
It's really hot out there.


``If`` statesments can be combined with loops (``for``, ``while``):

In [66]:
numbers = [-5, 3,2,-1,9,6]
total = 0
for n in numbers:
    if n >= 0:
        total += n
print(total)

20


In [67]:
def average(a):
    sum = 0.0
    for i in a:
        sum = sum + i
    return sum/len(a)

average([1,2,3,4])

2.5

In [69]:
def sumdif(x,y):
    sum, dif = x+y, x-y
    return sum, dif

a, b = sumdif(2,2)
print(a, b)

4 0


In [70]:
def main(n):
    cont = 0
    while (int(n) > 0):
        cont += 1
        n = n/2
#        print n
    return cont-1

main(10)
# main(10.3)

3

### Boolean operators.

In [71]:
a = 4
b = 40
(a>2) and (b>30)

True

In [72]:
(a>2) or (b>100)

True

In [73]:
not(a>2)

False

### Data Collections

We need to represent data collections: words in a text, students in a course, experimental data, etc., or to store intermediate results. The most simple data collection is the <code>list</code> (an ordered sequence of objects):

In [75]:
list(range(1,99,7))

[1, 8, 15, 22, 29, 36, 43, 50, 57, 64, 71, 78, 85, 92]

In [79]:
b = "This is an example".split()
print(b)

['This', 'is', 'an', 'example']


A list is an *ordered, mutable and dynamic* collection of *non-homogeneous* objects:

In [80]:
a = [1,2,3,4]
a[1] = 1
print(a)

[1, 1, 3, 4]


In [81]:
c = a + b
print(c)

[1, 1, 3, 4, 'This', 'is', 'an', 'example']


In [82]:
zeroes = [0] * 10
del zeroes[5:]
print(zeroes)

[0, 0, 0, 0, 0]


In [83]:
zeroes.append(1)
print(zeroes)

[0, 0, 0, 0, 0, 1]


In [84]:
zeroes.remove(1)
print(zeroes)

[0, 0, 0, 0, 0]


In [85]:
if 1 in zeroes:
    print(False)
else: 
    print(True)

True


### References

We can inspect the reference of an object:

In [86]:
a ='hello'
print(id(a))

140666847197592


Two different objects:

In [87]:
a = [1,2,3]
b = [1,2,3]
print(id(a), id(b))

140666837632008 140666837602504


Object alias:

In [88]:
a = [1,2,3]
b = a                     # alias
print(id(a), id(b))

140666830339272 140666830339272


Cloning:

In [89]:
a = [1,2,3]
b = a[:]                  # cloning with :

print (a, b, b[1:], id(a), id(b), id(b[1:]))

[1, 2, 3] [1, 2, 3] [2, 3] 140666830268360 140666830336520 140666830337480


When a list is an argument of a function, we are sending the *reference*, not a *copy*

In [90]:
def head(list):
    return list[0]

numbers=[1,2,3,4]
print(head(numbers), numbers)

1 [1, 2, 3, 4]


In [91]:
def change_first_element(list):
    list[0]=0
    
numbers=[1,2,3,4]
change_first_element(numbers)
print(numbers)

[0, 2, 3, 4]


If we return a list we are returning a reference:

In [92]:
def tail(list):
    return list[1:]     # we are creating a new list

numbers=[1,2,3,4]
rest = tail(numbers)
print(rest, numbers)
print(id(rest), id(numbers))

[2, 3, 4] [1, 2, 3, 4]
140666830311688 140666830313544


In [93]:
# Press tab
numbers.

SyntaxError: ignored

Sometimes it is important to perform a *sanity check* about what is doing a pre-defined function:

In [4]:
numbers=[1,2,3,4]

def test(l):
    return l.reverse()

# This method does not return any 
# value but reverse the given object from the list.


print(numbers) 
print(test(numbers), numbers)

[1, 2, 3, 4]
None [4, 3, 2, 1]


### Dictionaries

A dictionary is a collection that allows the access of an *element* by using a *key*:

In [5]:
dict = {"d": "D", "b":"B", "c":"C"}
dict["d"]

'D'

Dictionaries are *mutable* and *dynamic*:

In [6]:
dict["a"]="A"
type(dict)==dict

False

In [10]:
"a" in dict

True

In [12]:
del dict["a"]
print(dict)

{'d': 'D', 'b': 'B', 'c': 'C'}


In [13]:
dict = {"d": "D", "b":"B", "c":"C"}
dict.items()

dict_items([('d', 'D'), ('b', 'B'), ('c', 'C')])

### Tuples

Tuples are **non-mutable** lists:

In [14]:
tup = ('a', 'b', 'c')
print(type(tup), tup[1:3])

<class 'tuple'> ('b', 'c')


In [15]:
tup[0]='d'

TypeError: ignored

### Example: How to compute the statistics of words in a document.

In [18]:
# We can read a file in a list of line strings or as a string
text = open("https://raw.githubusercontent.com/DataScienceUB/DataScienceWorkshop-BS/master/text.txt",'r').readlines()
for l in text:
    print(l,)

FileNotFoundError: ignored

In [25]:
import urllib.request

file= 'https://raw.githubusercontent.com/DataScienceUB/DataScienceWorkshop-BS/master/text.txt'
page = urllib.request.urlopen(file)

for l in page:
    print(l,)

b'\r\n'
b'                 BORGES....... SU ULTIMO POEMA\r\n'
b'\r\n'
b'\tSi pudiera vivir nuevamente mi vida...\r\n'
b'\tEn la proxima cometeria mas errores.\r\n'
b'\tNo intentaria ser tan perfecto, \r\n'
b'\tme relajaria mas.\r\n'
b'\tSeria mas tonto de lo que he sido, \r\n'
b'\tde hecho tomaria muy pocas cosas con seriedad.\r\n'
b'\tSeria menos higienico.\r\n'
b'\tCorreria mas riesgos, \r\n'
b'\tharia mas viajes, \r\n'
b'\tcontemplaria mas atardeceres,\r\n'
b'\tsubiria mas montanas, \r\n'
b'\tnadaria mas rios.\r\n'
b'\tIria a mas lugares adonde nunca he ido, \r\n'
b'\tcomeria mas helados y menos  habas, \r\n'
b'\ttendria mas problemas reales \r\n'
b'\ty menos IMAGINARIOS.\r\n'
b'\tYo fui de esas personas que vivio \r\n'
b'\tsensata y prolificamente cada minuto de su vida; \r\n'
b'\tclaro que tuve alegrias.\r\n'
b'\tPero si pudiera volver atras trataria de tener\r\n'
b'\tsolamente buenos momentos.\r\n'
b'\tPor si no lo saben, \r\n'
b'\tde eso esta hecha la vida,\r\n'
b'\tsolo de mome

In [27]:
import requests
file= 'https://raw.githubusercontent.com/DataScienceUB/DataScienceWorkshop-BS/master/text.txt'
response = requests.get(file).text
response

'\r\n                 BORGES....... SU ULTIMO POEMA\r\n\r\n\tSi pudiera vivir nuevamente mi vida...\r\n\tEn la proxima cometeria mas errores.\r\n\tNo intentaria ser tan perfecto, \r\n\tme relajaria mas.\r\n\tSeria mas tonto de lo que he sido, \r\n\tde hecho tomaria muy pocas cosas con seriedad.\r\n\tSeria menos higienico.\r\n\tCorreria mas riesgos, \r\n\tharia mas viajes, \r\n\tcontemplaria mas atardeceres,\r\n\tsubiria mas montanas, \r\n\tnadaria mas rios.\r\n\tIria a mas lugares adonde nunca he ido, \r\n\tcomeria mas helados y menos  habas, \r\n\ttendria mas problemas reales \r\n\ty menos IMAGINARIOS.\r\n\tYo fui de esas personas que vivio \r\n\tsensata y prolificamente cada minuto de su vida; \r\n\tclaro que tuve alegrias.\r\n\tPero si pudiera volver atras trataria de tener\r\n\tsolamente buenos momentos.\r\n\tPor si no lo saben, \r\n\tde eso esta hecha la vida,\r\n\tsolo de momentos; \r\n\tno! no te pierdas el ahora.\r\n\tYo era uno de esos que nunca iban a ninguna parte\r\n\tsin u

In [0]:
# ipython magic commands

%ls

0.0 Jo.key
0.0.0.command*
0.1 Syllabus.key
0.2 What is Data Science.key
0.3 Python Data Science.key
1. Crash course on Python-NS.ipynb
1. Crash course on Python.ipynb*
6. Computational Statistics for Data Analysis.ipynb
7. Statistical Inference-Solutions.ipynb
7. Statistical Inference.html
7. Statistical Inference.ipynb
7. Statistical Inference.tex
7. Statistical Inference_files/
9. Bayesian Data Analysis (Copia en conflicto de MacBook Air de jordi 2015-11-05).ipynb
9. Bayesian Data Analysis-Solutions.ipynb
9. Bayesian Data Analysis.ipynb
9. Bayesian Data Analysis.ipynbfbqjl7/
Collaborative.ipynb
Regression_solutions.ipynb
Section4_3-Introduction-to-PyMC.ipynb
Test 1.ipynb
Untitled.ipynb
Untitled0.ipynb
files/
images/
marato/
movie_rating.csv
mu_1.png
mu_2.png
nu.png
pymc-2.3.6-cp27-none-win_amd64.whl
pymc.ipynb
python.rar
sigma_1.png
sigma_2.png
solution.py
styles/
test.csv
txtdata.csv


In [0]:
import string

def compare((w1,c1),(w2,c2)):
# A sorting funtion returns negative if x<y, zero if x==y, positive if x>y.
    if c1 < c2:
        return -1
    elif c1 == c2:
        return 0
    else:
        return 1
    
def main():
    
# We read the file in a string
    text = open("files/text.txt",'r').read()
    text = string.lower(text)
    for ch in '!"#$%&/()=?¿^*`+¨[]{}-:;,.':
        text = string.replace(text, ch, ' ')
# We build a list with all words (blank separation)
    words = string.split(text)

# We use a dictionary for counting words
    counts = {}
    for w in words:
        try:
            counts[w] = counts[w] + 1
        except KeyError:
            counts[w] = 1
            
# We create a sorted list
    items = counts.items()
    items.sort(compare, reverse=True)
    
# We print the list
    for i in range(len(items)):
        print "%-2s%3d" % items[i],'|' ,

main()    

mas 14 | de 10 | y   8 | que  6 | a   5 | si  5 | vida  4 | no  4 | menos  3 | la  3 | pudiera  3 | pero  3 | me  3 | un  3 | estoy  2 | an~os  2 | lo  2 | ven  2 | yo  2 | ya  2 | volver  2 | el  2 | momentos  2 | en  2 | contemplaria  2 | 85  2 | nunca  2 | con  2 | por  2 | vivir  2 | seria  2 | he  2 | tengo  2 | muriendo  2 | su  2 | se  2 | montanas  1 | tener  1 | paracaidas  1 | oton~o  1 | calesita  1 | borges  1 | otra  1 | comeria  1 | iban  1 | personas  1 | te  1 | atardeceres  1 | termometro  1 | sensata  1 | buenos  1 | vueltas  1 | agua  1 | perfecto  1 | andar  1 | haria  1 | bolsa  1 | una  1 | cosas  1 | ninguna  1 | tendria  1 | sido  1 | ser  1 | esos  1 | pocas  1 | tuve  1 | muy  1 | errores  1 | riesgos  1 | jugaria  1 | eso  1 | solo  1 | nadaria  1 | poema  1 | descalzo  1 | problemas  1 | prolificamente  1 | asi  1 | pierdas  1 | subiria  1 | ahora  1 | ido  1 | iria  1 | tomaria  1 | claro  1 | helados  1 | intentaria  1 | cada  1 | amaneceres  1 | relajaria

### Lists (and dictionary) comprehensions

Lists comprehensions are a way to fit a ``for`` loop, an ``if`` statement, and an assignment all in one line.

A list comprehension consists of the following parts:

+ An input sequence.
+ A variable representing members of the input sequence.
+ An optional expression.
+ An output expression producing elements of the output list from members of the input sequence that satisfy the predicate.

In [0]:
num = [1, 4, -5, 10, -7, 2, 3, -1]
squared = [ x**2 for x in num if x > 0]
print type(squared), squared

<type 'list'> [1, 16, 100, 4, 9]


There is a downside to list comprehensions: the entire list has to be stored in memory at once. This isn’t a problem for small lists like the ones in the above examples, or even of lists several orders of magnitude larger. But we can use **<font color="red">generators</font>** to solve this problem.

Generator expressions do not load the whole list into memory at once, but instead create a *generator object* so only one list element has to be loaded at any time.

Generator expressions have the same syntax as list comprehensions, but with parentheses around the outside instead of brackets:

In [0]:
num = [1, 4, -5, 10, -7, 2, 3, -1]
squared = ( x**2 for x in num if x > 0 )
print type(squared), squared

<type 'generator'> <generator object <genexpr> at 0x11149dcd0>


The elements of the generator must be accessed by an iterator because they are generated when needed:

In [0]:
lis = []
for item in squared:
    lis = lis + [item]
print lis

[1, 16, 100, 4, 9]


We can define our own generators with the ``yield`` statesment. For example, let's build a generator for the binary representation of a number between 0 and 1 with arbitrary precision.

In [0]:
# binary representation of a number between 0 and 1 (b bits precision).

def res(n,b):
    bin_a = '.'
    for i in range(b):
        n *= 2
        bin_a +=  str(int(n))
        n = n % 1
    return bin_a

print res(1/3.0,10)

In [0]:
# binary representation of a number between 0 and 1 (precision as needed).

def binRep(n):
    while True:
        n *= 2
        yield int(n)
        n = n % 1

a = binRep(1/3.) 
a_bin = '.'
for i in range(50):
    a_bin +=  str(a.next())
    
print a_bin

** Exercise **

+ Write a function with two parameters <code>a</code> and <code>b</code>, to compute the final amount we get if we deposite 1000€ during <code>a</code> years in a bank account with an interest rate of <code>b</code> per cent.
+ What is the result for ``a``=10, ``b``=10

In [0]:
# Your solution here

In [0]:
# My solution

def amount(a,b):
    am = 1000.0
    for i in range(a):
        am = am + am* b/100.
    return am

amount(10,10)

**Exercise**

+ Write a function with one parameter <code>a</code>, to compute the minimum period we need to double the amount in an account with an interest rate of <code>a</code> per cent.
+ What is the result for ``a``=3?

In [0]:
# Your solution here

In [0]:
# My solution

def double(a):
    amount = 1.0
    c = 1
    while amount<2.0:
        amount += amount * a/100.
        c += 1
    return c-1

double(3)

** Exercise**

> (...) In mathematics, the sieve of Eratosthenes, one of a number of prime number sieves, is a simple, ancient algorithm for finding all prime numbers up to any given limit. A prime number is a natural number which has exactly two distinct natural number divisors: 1 and itself. To find all the prime numbers less than or equal to a given integer $n$ by Eratosthenes' method.
> (Source: *Wikipedia*)

+ Write a program to implement the Eratostenes algorithm. First create a list of consecutive integers from 2 through $n$: (2, 3, 4, ..., n). Then,    
    - Initially, let $p$ equal 2, the first prime number.
    - Starting from $p$, enumerate its multiples by counting to $n$ in increments of $p$, and mark them in the list (these will be 2p, 3p, 4p, etc.; the $p$ itself should not be marked).
    - Find the first number greater than $p$ in the list that is not marked. If there was no such number, stop. Otherwise, let $p$ now equal this new number (which is the next prime), and repeat from step 2.
    - When the algorithm terminates, all the numbers in the list that are not marked are prime. (...)

In [0]:
# Your solution here

In [0]:

# My solution
def eras(n):
  siv=range(n+1)
  # 1 is not prime
  siv[1]=0
  # The last possible factor is sqrt(n)
  sqn=int(round(n**0.5))
  for i in range(2,sqn+1):
    if siv[i]!=0:
        siv[2*i:n/i*i+1:i]=[0]*(n/i-1)
  return filter(None,siv)
    
print eras(100000)

%timeit eras(10000)

[2, 3, 5, 7, 11, 13, 17, 19, 23, 29, 31, 37, 41, 43, 47, 53, 59, 61, 67, 71, 73, 79, 83, 89, 97, 101, 103, 107, 109, 113, 127, 131, 137, 139, 149, 151, 157, 163, 167, 173, 179, 181, 191, 193, 197, 199, 211, 223, 227, 229, 233, 239, 241, 251, 257, 263, 269, 271, 277, 281, 283, 293, 307, 311, 313, 317, 331, 337, 347, 349, 353, 359, 367, 373, 379, 383, 389, 397, 401, 409, 419, 421, 431, 433, 439, 443, 449, 457, 461, 463, 467, 479, 487, 491, 499]
1000 loops, best of 3: 495 µs per loop


**Exercise:** 

+ Compute the set of bigrams of a string. (``'hola'->'ho'+'ol'+'la'``) 

In [0]:
# Your solution here

In [0]:
%%writefile -a solution.py

# My solution

def bigrams(a):
    s = set()
    for i in range(len(a)-1):
        s.add(a[i:i+2])
    return s

print bigrams("llangardaix")
print bigrams("setze jutges d'un jutjat")

In [0]:
# %load solution.py

# My solution

def bigrams(a):
    s = set()
    for i in range(len(a)-1):
        s.add(a[i:i+2])
    return s

print bigrams("llangardaix")
print bigrams("setze jutges d'un jutjat")

### Objects

You can define your own classes and objects.

In [0]:
#creating a class

class Rectangle:
    def __init__(self,x,y):
        self.x = x
        self.y = y
    description = "This shape has not been described yet"
    author = "Nobody has claimed to make this shape yet"
    def area(self):
        return self.x * self.y
    def perimeter(self):
        return 2 * self.x + 2 * self.y
    def describe(self,text):
        self.description = text
    def authorName(self,text):
        self.author = text
    def scaleSize(self,scale):
        self.x = self.x * scale
        self.y = self.y * scale

#creating objects
a = Rectangle(100, 45)
b = Rectangle(10,230)

#describing the rectangles
a.describe("A fat rectangle")
b.describe("A thin rectangle")

In [0]:
#finding the area of your rectangle:
print a.area()
 
#finding the perimeter of your rectangle:
print a.perimeter()

#getting the description
print a.description
print a.author

In [0]:
#finding the area of your rectangle:
print b.area()
print b.description

#making the rectangle 50% smaller
b.scaleSize(0.5)
b.describe("A small thin rectangle")
 
#re-printing the new area of the rectangle
print b.area()
print b.description

## Reading Data

The most convenient method that you can use to work with data is to load it directly into memory.

When you load a file (of any type), the entire dataset is available at all times and the loading process is quite direct:

In [0]:
with open("files/SeaIce.txt", 'r') as input_file:
    print 'File content', input_file.read()

File content year mo    data_type region extent   area
1979  1      Goddard      N  15.54  12.33
1980  1      Goddard      N  14.96  11.85
1981  1      Goddard      N  15.03  11.82
1982  1      Goddard      N  15.26  12.11
1983  1      Goddard      N  15.10  11.92
1984  1      Goddard      N  14.61  11.60
1985  1      Goddard      N  14.86  11.60
1986  1      Goddard      N  15.02  11.79
1987  1      Goddard      N  15.20  11.81
1988  1        -9999      N  -9999  -9999
1989  1      Goddard      N  15.12  13.11
1990  1      Goddard      N  14.95  12.72
1991  1      Goddard      N  14.46  12.49
1992  1      Goddard      N  14.72  12.54
1993  1      Goddard      N  15.08  12.85
1994  1      Goddard      N  14.82  12.80
1995  1      Goddard      N  14.62  12.72
1996  1      Goddard      N  14.21  12.07
1997  1      Goddard      N  14.47  12.30
1998  1      Goddard      N  14.81  12.73
1999  1      Goddard      N  14.47  12.54
2000  1      Goddard      N  14.41  12.22
2001  1      Goddard 

The entire dataset is loaded from the library into free memory. Of course, the loading process will fail if your system lacks sufficient memory to hold the dataset. When this problem occurs, you need to consider other techniques
for working with the dataset, such as **streaming** it or **sampling** it.

Here’s an example of how you can stream data using Python:

In [0]:
with open("files/SeaIce.txt", 'r') as input_file:
    for observation in input_file:
        print 'Reading Data: ' + observation,

The ``input_file`` file object contains a pointer to the open file. As the code performs data reads in the for loop, the file pointer moves to the next record.

Data streaming obtains all the records from a data source. You may find that
you don’t need all the records. You can save time and resources by simply
sampling the data.

In [0]:
n = 17
with open("files/SeaIce.txt", 'r') as input_file:
    for j, observation in enumerate(input_file):
        if j % n==0:
            print('Reading Line: ' + str(j) + ' Content: ' + observation), 

You can perform random sampling as well.

In [0]:
import random
sample_size = 0.01
with open("files/SeaIce.txt", 'r') as input_file:
    for j, observation in enumerate(input_file):
        if random.random()<=sample_size:
            print('Reading Line: ' + str(j) + ' Content: ' + observation), 

A flat file presents the easiest kind of file to work with. 

A problem with using native Python techniques is that the input isn’t intelligent. For example, when a file contains a header, Python simply reads it as yet more data to process, rather than as a header (not a problem for Pandas!).

The least formatted and therefore easiest‐to‐read flat‐file format is the text file. However, a text file also treats all data as strings, so you often have to convert numeric data into other forms.

A comma‐separated value (CSV) file provides more formatting and more information, but it requires a little more effort to read.

At the high end of flat‐file formatting are custom data formats, such as an Excel file, which contains extensive formatting and could include multiple datasets in a single file.

### Reading from a CSV file

A CSV file provides more formatting than a simple text file. In fact, CSV files can become quite complicated. There is a standard that defines the format of CSV files, and you can see it at https://tools.ietf.org/html/rfc4180.

The ``csv`` module is useful for working with data exported from spreadsheets and databases into text files formatted with fields and records, commonly referred to as comma-separated value (CSV).

In [0]:
import csv

f = open("files/Advertising.csv", 'r')
try:
    reader = csv.reader(f)
    for row in reader:
        print row
finally:
    f.close()

In [0]:
with open("files/Advertising.csv", 'r') as input_file:
    reader = csv.reader(input_file)
    for row in reader:
        print row

When you have data to be imported into some other application, writing ``csv`` files is just as easy as reading them. 

Use ``writer()`` to create an object for writing, then iterate over the rows, using ``writerow()`` to print them.  

In [0]:
import csv

ifile  = open("files/Advertising.csv", 'r')
reader = csv.reader(ifile)

ofile  = open('test.csv', "w")
writer = csv.writer(ofile, delimiter=',', lineterminator='\n')

for row in reader:
    writer.writerow(row)

ifile.close()
ofile.close()

A more ellegant way of reading a CSV file:
+  Wrap the CSV reader in a function that returns a generator
+  Use context managers ``with [callable] as [name]`` to ensure that the handle to the file is closed automatically.
+  Use the ``csv.DictReader`` class when headers are present (otherwise just use ``csv.reader``)

In [0]:
import csv

ADV = 'files/Advertising.csv'

def read_data(path):
    with open(path, 'r') as data:
        reader = csv.DictReader(data)
        for row in reader:
            yield row

for idx, row in enumerate(read_data(ADV)):
    if idx < 10: print row
    else: break

The file is not opened, read, or parsed until you need it. This is powerful because it means that even for much larger data sets you will have efficient, portable code. 

In [0]:
data = read_data(ADV)
print data

### Some goodies

In [0]:
!pip install tqdm 

In [0]:
import tqdm
from time import sleep
text = ""
for char in tqdm.tqdm(range(1000)):
    sleep(0.01)